rskj-core/src/main/java/co/rsk/peg/BridgeUtils.java

```diff
@@ -243,4 +241,49 @@ private static boolean isFromFeePerKbChangeAuthorizedSender(org.ethereum.core.Tr
         AddressBasedAuthorizer authorizer = bridgeConfiguration.getFeePerKbChangeAuthorizer();
         return authorizer.isAuthorized(rskTx);
     }
+
+    public static boolean validateHeightAndConfirmations(int height, int btcBestChainHeight, int acceptableConfirmationsAmount, Sha256Hash btcTxHash) throws Exception {
+        // Check there are at least N blocks on top of the supplied height
+        if (height < 0) {
+            throw new Exception();
+        }
+        int confirmations = btcBestChainHeight - height + 1;
+        if (confirmations < acceptableConfirmationsAmount) {
+            logger.warn(
+                    "Btc Tx {} at least {} confirmations are required, but there are only {} confirmations",
+                    btcTxHash,
+                    acceptableConfirmationsAmount,
+                    confirmations
+            );
+            return false;
+        }
+        return true;
+    }
+
+    public static Sha256Hash calculateMerkleRoot(NetworkParameters networkParameters, byte[] pmtSerialized, Sha256Hash btcTxHash) throws VerificationException{
+        PartialMerkleTree pmt = new PartialMerkleTree(networkParameters, pmtSerialized, 0);
+        List<Sha256Hash> hashesInPmt = new ArrayList<>();
+        Sha256Hash merkleRoot = pmt.getTxnHashAndMerkleRoot(hashesInPmt);
+        if (!hashesInPmt.contains(btcTxHash)) {
+            logger.warn("Supplied Btc Tx {} is not in the supplied partial merkle tree", btcTxHash);
+            return null;
+        }
+        return merkleRoot;
+    }
+
+    public static void validateInputsCount(byte[] btcTxSerialized, boolean isActiveRskip, Sha256Hash btcTxHash) throws VerificationException.EmptyInputsOrOutputs {
```

Comment:
You could probably remove hash parameter as it doesn't have an actual usage internally, and instead, logging in the callers directly.